plugins/elevenlabs/src/tts.ts

```diff
@@ -0,0 +1,259 @@
+import { type RawData, WebSocket } from 'ws';
+import type { TTSEncoding, TTSModels } from './models.js';
+
+type Voice = {
+  id: string;
+  name: string;
+  category: string;
+  settings?: VoiceSettings;
+};
+
+type VoiceSettings = {
+  stability: number; // 0..1
+  similarity_boost: number; // 0..1
+  style?: number; // 0..1
+  use_speaker_boost: boolean;
+};
+
+const DEFAULT_VOICE: Voice = {
+  id: 'EXAVITQu4vr4xnSDxMaL',
+  name: 'Bella',
+  category: 'premade',
+  settings: {
+    stability: 0.71,
+    similarity_boost: 0.5,
+    style: 0.0,
+    use_speaker_boost: true,
+  },
+};
+
+const API_BASE_URL_V1 = 'https://api.elevenlabs.io/v1';
+const AUTHORIZATION_HEADER = 'xi-api-key';
+
+export interface TTSOptions {
+  apiKey?: string;
+  voice: Voice;
+  modelID: TTSModels;
+  baseURL: string;
+  encoding: TTSEncoding;
+  streamingLatency: number;
+  wordTokenizer: tokenize.WordTokenizer;
+  chunkLengthSchedule: number[];
+  enableSsmlParsing: boolean;
+}
+
+const defaultTTSOptions: TTSOptions = {
+  apiKey: process.env.ELEVEN_API_KEY,
+  voice: DEFAULT_VOICE,
+  modelID: 'eleven_turbo_v2_5',
+  baseURL: API_BASE_URL_V1,
+  encoding: 'pcm_22050',
+  streamingLatency: 3,
+  wordTokenizer: new tokenize.basic.WordTokenizer(false),
+  chunkLengthSchedule: [],
+  enableSsmlParsing: false,
+};
+
+export class TTS extends tts.TTS {
+  #opts: TTSOptions;
+
+  constructor(opts = defaultTTSOptions) {
+    super(sampleRateFromFormat(opts.encoding), 1, { streaming: true });
+    if (opts.apiKey === undefined) {
+      throw new Error(
+        'ElevenLabs API key is required, whether as an argument or as $ELEVEN_API_KEY',
+      );
+    }
+
+    this.#opts = { ...defaultTTSOptions, ...opts };
+  }
+
+  async listVoices(): Promise<Voice[]> {
+    return fetch(this.#opts.baseURL + '/voices', {
+      headers: {
+        [AUTHORIZATION_HEADER]: this.#opts.apiKey!,
+      },
+    })
+      .then((data) => data.json())
+      .then((data) => {
+        const voices: Voice[] = [];
+        for (const voice of (
+          data as { voices: { voice_id: string; name: string; category: string }[] }
+        ).voices) {
+          voices.push({
+            id: voice.voice_id,
+            name: voice.name,
+            category: voice.category,
+            settings: undefined,
+          });
+        }
+        return voices;
+      });
+  }
+
+  stream(): tts.SynthesizeStream {
+    return new SynthesizeStream(this.#opts);
+  }
+}
+
+export class SynthesizeStream extends tts.SynthesizeStream {
+  closed: boolean;
+  #opts: TTSOptions;
+  #logger = log();
+  readonly streamURL: string;
+
+  constructor(opts: TTSOptions) {
+    super();
+    this.#opts = opts;
+    this.closed = false;
+    this.streamURL = `${this.#opts.baseURL}/text-to-speech/${this.#opts.voice.id}/stream-input?model_id=${this.#opts.modelID}&output_format=${this.#opts.encoding}&optimize_streaming_latency=${this.#opts.streamingLatency}&enable_ssml_parsing=${this.#opts.enableSsmlParsing}`;
```

Comment:
can we either `encodeURIComponent` or use `URLSearchParams` or other helpers to construct this URL in a safe way?